cupyx/scipy/linalg/decomp_lu.py

```diff
@@ -0,0 +1,174 @@
+import numpy
+
+import cupy
+from cupy import cuda
+from cupy.cuda import cublas
+from cupy.cuda import device
+from cupy.linalg import util
+
+if cuda.cusolver_enabled:
+    from cupy.cuda import cusolver
+
+
+def lu_factor(a, overwrite_a=False, check_finite=True):
```

Comment:
Thanks. I didn't noticed that.

```
In [12]: cupyx.scipy.linalg.lu_factor(cp.array([[0,1],[0,0]], dtype=cp.float32))
Out[12]: 
(array([[ 0.,  1.],
        [nan, nan]], dtype=float32), array([0, 1], dtype=int32))

In [13]: scipy.linalg.lu_factor(np.array([[0,1],[0,0]], dtype=np.float32))
/home/sakai/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/scipy/linalg/decomp_lu.py:71: RuntimeWarning: Diagonal number 1 is exactly zero. Singular matrix.
  RuntimeWarning)
Out[13]: 
(array([[0., 1.],
        [0., 0.]], dtype=float32), array([0, 1], dtype=int32))
```

Hmm, it would be possible to provide the same warning based on  the `devInfo` result from cuSolver.

But I think changing `nan` into `0` is not worth paying the cost.
Is it better to document the difference?

Reply:
OK, let's descript it in the documentation.
How about enhancing the error messege in `lu_solve` in the case of `check_finite=True`?

Reply:
I added a note describing the difference.

> How about enhancing the error messege in `lu_solve` in the case of `check_finite=True`?

Do you mean that the error message should indicate the error might be caused by this difference of behavior?

Reply:
Sorry. I'd like to reconsider whether we convert `non` to `0` because I realized that we can perform such conversion only when `dev_info[0] > 0`.

Reply:
Finally I gave up converting `nan` to `0.`.

In the following case, we need to convert both `nan` to `0.`.

```
>>> scipy.linalg.lu_factor(np.array([[0,1],[0,0]], dtype=np.float32))
__main__:1: LinAlgWarning: Diagonal number 1 is exactly zero. Singular matrix.
(array([[0., 1.],
       [0., 0.]], dtype=float32), array([0, 1], dtype=int32))
>>> cupyx.scipy.linalg.lu_factor(cp.array([[0,1],[0,0]], dtype=cp.float32))
__main__:1: RuntimeWarning: Diagonal number 1 is exactly zero. Singular matrix.
(array([[ 0.,  1.],
       [nan, nan]], dtype=float32), array([0, 1], dtype=int32))
```

But in the following case, we need to convert only one of `nan` to `0.`.

```
>>> nan = float("nan")
>>> scipy.linalg.lu_factor(np.array([[0,1],[0,nan]], dtype=np.float32), check_finite=False)
__main__:1: LinAlgWarning: Diagonal number 1 is exactly zero. Singular matrix.
(array([[ 0.,  1.],
       [ 0., nan]], dtype=float32), array([0, 1], dtype=int32))
>>> cupyx.scipy.linalg.lu_factor(cp.array([[0,1],[0,nan]], dtype=cp.float32), check_finite=False)
__main__:1: RuntimeWarning: Diagonal number 1 is exactly zero. Singular matrix.
(array([[ 0.,  1.],
       [nan, nan]], dtype=float32), array([0, 1], dtype=int32))
```

It seems impossible (or at least difficult without knowing details of the two `getrf` implementations) for me to perform correct conversion.

Reply:
I enhanced exception message of `lu_solve`.